File: diagramas/ChefApp_v2/ChefAppClean/app/src/main/java/com/chefapp/fragments/WebFragment.java
package com.chefapp.fragments;

import android.os.Bundle;
import android.view.KeyEvent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.inputmethod.EditorInfo;
import android.webkit.WebChromeClient;
import android.webkit.WebSettings;
import android.webkit.WebView;
import android.webkit.WebViewClient;
import android.widget.EditText;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.chefapp.R;
import com.chefapp.adapters.QuickAccessAdapter;
import com.chefapp.models.DataProvider;

public class WebFragment extends Fragment {

    private WebView webView;
    private EditText etUrl;
    private TextView tvStatus;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater,
                             @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_web, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        webView = view.findViewById(R.id.webview);
        etUrl = view.findViewById(R.id.et_url);
        tvStatus = view.findViewById(R.id.tv_status);

        // Configure WebView
        WebSettings settings = webView.getSettings();
        settings.setJavaScriptEnabled(true);
        settings.setDomStorageEnabled(true);
        settings.setLoadWithOverviewMode(true);
        settings.setUseWideViewPort(true);

        webView.setWebViewClient(new WebViewClient() {
            @Override public void onPageFinished(WebView view, String url) {
                tvStatus.setText("✓ Listo");
                tvStatus.setTextColor(getResources().getColor(R.color.imc_normal, null));
                etUrl.setText(url);
            }
            @Override public void onPageStarted(WebView view, String url, android.graphics.Bitmap favicon) {
                tvStatus.setText("⟳ Cargando...");
                tvStatus.setTextColor(getResources().getColor(R.color.text_secondary, null));
            }
        });

        webView.setWebChromeClient(new WebChromeClient());

        // Go button
        view.findViewById(R.id.btn_go).setOnClickListener(v -> loadUrl());

        // URL field Enter
        etUrl.setOnEditorActionListener((v, actionId, event) -> {
            if (actionId == EditorInfo.IME_ACTION_GO ||
                    (event != null && event.getKeyCode() == KeyEvent.KEYCODE_ENTER)) {
                loadUrl();
                return true;
            }
            return false;
        });

        // Nav buttons
        view.findViewById(R.id.btn_back).setOnClickListener(v -> {
            if (webView.canGoBack()) webView.goBack();
        });
        view.findViewById(R.id.btn_forward).setOnClickListener(v -> {
            if (webView.canGoForward()) webView.goForward();
        });
        view.findViewById(R.id.btn_reload).setOnClickListener(v -> webView.reload());

        // Quick Access RecyclerView
        RecyclerView rvQuick = view.findViewById(R.id.rv_quick_access);
        rvQuick.setLayoutManager(new LinearLayoutManager(getContext()));
        QuickAccessAdapter adapter = new QuickAccessAdapter(DataProvider.getQuickAccessSites());
        rvQuick.setAdapter(adapter);
        adapter.setOnSiteClickListener(site -> {
            etUrl.setText(site.getUrl());
            loadUrlString(site.getUrl());
        });

        // Load default URL
        loadUrlString("https://www.recetasgratis.net");
    }

    private void loadUrl() {
        String url = etUrl.getText().toString().trim();
        if (!url.isEmpty()) loadUrlString(url);
    }

    private void loadUrlString(String url) {
        if (!url.startsWith("http://") && !url.startsWith("https://")) {
            url = "https://" + url;
        }
        webView.loadUrl(url);
    }

    @Override
    public void onDestroyView() {
        if (webView != null) {
            webView.destroy();
            webView = null;
        }
        super.onDestroyView();
    }
}

File: app/src/main/java/com/example/chefapp/fragments/WebFragment.kt
package com.example.chefapp.fragments

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.webkit.WebChromeClient
import android.webkit.WebView
import android.webkit.WebViewClient
import androidx.fragment.app.Fragment
import com.example.chefapp.databinding.FragmentWebBinding
import com.example.chefapp.helpers.NavegadorWeb

class WebFragment : Fragment() {

    private var _binding: FragmentWebBinding? = null
    private val binding get() = _binding!!

    private val navegador = NavegadorWeb()

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        _binding = FragmentWebBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        configurarWebView()

        binding.btnIr.setOnClickListener {
            val url = binding.etUrl.text.toString()
            if (url.isNotEmpty()) cargarUrl(url)
        }

        binding.btnAtras.setOnClickListener {
            if (binding.webView.canGoBack()) binding.webView.goBack()
        }

        binding.btnAdelante.setOnClickListener {
            if (binding.webView.canGoForward()) binding.webView.goForward()
        }

        binding.btnRecargar.setOnClickListener {
            binding.webView.reload()
        }

        val accesos = navegador.accesosRapidos
        binding.itemAcceso1.setOnClickListener { cargarUrl(accesos[0].second) }
        binding.itemAcceso2.setOnClickListener { cargarUrl(accesos[1].second) }
        binding.itemAcceso3.setOnClickListener { cargarUrl(accesos[2].second) }
        binding.itemAcceso4.setOnClickListener { cargarUrl(accesos[3].second) }

        binding.tvAcceso1.text = accesos[0].first
        binding.tvAcceso2.text = accesos[1].first
        binding.tvAcceso3.text = accesos[2].first
        binding.tvAcceso4.text = accesos[3].first

        cargarUrl(navegador.urlActual)
    }

    private fun configurarWebView() {
        binding.webView.apply {
            settings.javaScriptEnabled = true
            settings.domStorageEnabled = true
            settings.loadWithOverviewMode = true
            settings.useWideViewPort = true

            webViewClient = object : WebViewClient() {
                override fun onPageStarted(view: WebView?, url: String?, favicon: android.graphics.Bitmap?) {
                    binding.tvEstado.text = "● Cargando"
                    binding.progressWebView.visibility = View.VISIBLE
                    binding.etUrl.setText(url)
                }
                override fun onPageFinished(view: WebView?, url: String?) {
                    binding.tvEstado.text = "✓ Listo"
                    binding.progressWebView.visibility = View.GONE
                }
            }

            webChromeClient = object : WebChromeClient() {
                override fun onProgressChanged(view: WebView?, newProgress: Int) {
                    binding.progressWebView.progress = newProgress
                }
            }
        }
    }

    private fun mostrarHistorial() {
        actualizarHistorialUI()
    }

    private fun cargarUrl(url: String) {
        navegador.cargarPagina(url)
        binding.etUrl.setText(navegador.urlActual)
        binding.webView.loadUrl(navegador.urlActual)
        actualizarHistorialUI()
    }

    fun agregarAccesoRapido(nombre: String, url: String) {
        navegador.agregarAccesoRapido(nombre, url)
    }

    private fun actualizarHistorialUI() {
        val historial = navegador.mostrarHistorial()
        val tvs = listOf(
            binding.tvHistorial1, binding.tvHistorial2,
            binding.tvHistorial3, binding.tvHistorial4, binding.tvHistorial5
        )
        val rows = listOf(
            binding.rowHistorial1, binding.rowHistorial2,
            binding.rowHistorial3, binding.rowHistorial4, binding.rowHistorial5
        )
        tvs.forEachIndexed { i, tv ->
            if (i < historial.size) {
                rows[i].visibility = View.VISIBLE
                tv.text = "🕓 ${historial[i]}"
                rows[i].setOnClickListener { cargarUrl(historial[i]) }
            } else {
                rows[i].visibility = View.GONE
            }
        }
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }
}